JobRunner/SpecialRunner.py

```diff
@@ -0,0 +1,141 @@
+import os
+from threading import Thread
+from subprocess import Popen, PIPE
+from time import sleep
+from select import select
+
+
+class SpecialRunner:
+    """
+    This class provides the interface to run special job types like
+    CWL, WDL and HPC.
+
+    """
+
+    def __init__(self, config, job_id, logger=None):
+        """
+        Inputs: config dictionary, Job ID, and optional logger
+        """
+        self.config = config
+        self.top_job_id = job_id
+        self.logger = logger
+        self.token = config['token']
+        self.workdir = config.get('workdir', '/mnt/awe/condor')
+        self.shareddir = os.path.join(self.workdir, 'workdir/tmp')
+        self.containers = []
+        self.threads = []
+        self.allowed = ['slurm']
+
+    _POLL = 0.1
+    _POLL2 = 0.1
+
+    def run(self, config, data, job_id, callback=None, fin_q=[]):
+        # TODO:
+        # initialize working space
+        # check job type against an allow list
+        # submit the job and map the batch jobb to the job id
+        # start a thread to monitor progress
+        (module, method) = data['method'].split('.')
+
+        if module != 'special':
+            raise ValueError("Incorrect module should be 'special'")
+
+        if method not in self.allowed:
+            raise ValueError("Invalid special method type")
+
+        return self._batch_submit(method, config, data, job_id, fin_q)
+
+    def _check_batch_job(self, check, slurm_jobid):
+        cmd = [check, slurm_jobid]
```

Comment:
I wonder if its better to use pySlurm in the future.

Reply:
I haven't heard good things about it.  I don't think it works too well.